File: mobile/android/push/src/main/java/boo/gcoms/push/PushGateway.kt
package boo.gcoms.push

import boo.gcoms.sdk.GComs
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONArray
import org.json.JSONObject
import java.net.URL
import javax.net.ssl.HttpsURLConnection

/** Application-owned HTTPS endpoint; app authentication issues tickets separately. */
class PushGateway(origin: String, private val storage: PushStorage) {
    private val base = URL(origin)
    private val mutex = Mutex()
    init {
        require(base.protocol == "https" && base.host.isNotEmpty() && base.userInfo == null &&
            base.query == null && base.ref == null && base.path in listOf("", "/"))
    }
    private suspend fun post(path: String, body: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        val bytes = body.toString().toByteArray(Charsets.UTF_8)
        require(bytes.size <= 8192)
        val connection = URL(base, path).openConnection() as HttpsURLConnection
        try {
            connection.instanceFollowRedirects = false
            connection.connectTimeout = 10000
            connection.readTimeout = 10000
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.doOutput = true
            connection.setFixedLengthStreamingMode(bytes.size)
            connection.outputStream.use { it.write(bytes) }
            check(connection.responseCode in 200..299) { "Push gateway rejected request" }
            connection.inputStream.use {
                val output = java.io.ByteArrayOutputStream()
                val buffer = ByteArray(1024)
                while (true) {
                    val count = it.read(buffer)
                    if (count < 0) break
                    check(output.size() + count <= 8192) { "Push gateway response is too large" }
                    output.write(buffer, 0, count)
                }
                val reply = output.toByteArray()
                JSONObject(reply.toString(Charsets.UTF_8))
            }
        } finally { bytes.fill(0); connection.disconnect() }
    }
    /** Token rotation needs a fresh one-use ticket from the app's authenticated server. */
    suspend fun register(ticket: String, deviceToken: String) = mutex.withLock {
        require(ticket.length <= 4096 && deviceToken.length in 1..4096)
        val reply = post("/v1/register", JSONObject().put("ticket", ticket)
            .put("platform", "fcm").put("token", deviceToken))
        val reference = reply.getString("reference")
        val management = reply.getString("management_token")
        require(reference.matches(Regex("[0-9a-f]{64}")) && management.matches(Regex("[0-9a-f]{64}")))
        val state = storage.load()
        state.put("reference", reference).put("management_token", management).put("expires", reply.getLong("expires"))
        storage.save(state)
    }
    /** Persist a revision before binding; call after reopen and before suspension. */
    suspend fun bind(sdk: GComs) = mutex.withLock {
        val state = storage.load()
        val reference = state.optString("reference", "0".repeat(64))
        val revision = Math.addExact(state.optLong("revision", 0), 1)
        val now = System.currentTimeMillis() / 1000
        val expiry = if (reference == "0".repeat(64)) now + 3600 else minOf(now + 23 * 3600, state.getLong("expires"))
        require(expiry > now) { "Refresh expired push registration first" }
        state.put("revision", revision)
        storage.save(state)
        val bytes = JSONArray((0 until 32).map { reference.substring(it * 2, it * 2 + 2).toInt(16) })
        sdk.request(JSONObject().put("op", "bind_push").put("reference", bytes)
            .put("revision", revision).put("expires", expiry))
    }
    /** Provider opt-out first. Call bind afterwards to remove current relay bindings. */
    suspend fun unregister() = mutex.withLock {
        val state = storage.load()
        if (state.has("reference")) post("/v1/unregister", JSONObject()
            .put("reference", state.getString("reference")).put("management_token", state.getString("management_token")))
        storage.save(JSONObject().put("revision", state.optLong("revision", 0)))
    }
}

File: mobile/apple/Sources/GComsPush/PushGateway.swift
import Foundation
import Security
import GComs

public struct PushState: Codable, Sendable {
    public var reference: String?
    public var managementToken: String?
    public var expires: UInt64 = 0
    public var revision: UInt64 = 0
    public init() {}
}

public protocol PushStorage: Sendable {
    func load() async throws -> PushState
    func save(_ value: PushState) async throws
}

/// Use one gateway per profile. Registration tickets come from the app's server.
public actor PushGateway {
    private let origin: URL
    private let storage: any PushStorage
    private let session: URLSession
    private var busy = false
    public init(origin: URL, storage: any PushStorage) throws {
        guard origin.scheme == "https", origin.host != nil, origin.user == nil,
            origin.password == nil, origin.query == nil, origin.fragment == nil,
            ["", "/"].contains(origin.path) else { throw GComsError.invalidResponse }
        self.origin = origin
        self.storage = storage
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 10
        config.httpMaximumConnectionsPerHost = 1
        self.session = URLSession(configuration: config, delegate: NoRedirect(), delegateQueue: nil)
    }
    private func post(_ path: String, _ body: [String: Any]) async throws -> [String: Any] {
        let data = try JSONSerialization.data(withJSONObject: body)
        guard data.count <= 8192 else { throw GComsError.invalidResponse }
        var request = URLRequest(url: origin.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = data
        let (stream, response) = try await session.bytes(for: request)
        guard let response = response as? HTTPURLResponse, (200..<300).contains(response.statusCode) else {
            throw GComsError.native("Push gateway rejected request")
        }
        var received = Data()
        for try await byte in stream {
            guard received.count < 8192 else { throw GComsError.invalidResponse }
            received.append(byte)
        }
        guard let result = try JSONSerialization.jsonObject(with: received) as? [String: Any] else {
            throw GComsError.invalidResponse
        }
        return result
    }
    /// Call from didRegisterForRemoteNotificationsWithDeviceToken, with a fresh ticket.
    public func register(ticket: String, deviceToken: Data) async throws {
        guard !busy else { throw GComsError.unavailable }
        busy = true
        defer { busy = false }
        guard ticket.utf8.count <= 4096, !deviceToken.isEmpty, deviceToken.count <= 256 else {
            throw GComsError.invalidResponse
        }
        let token = deviceToken.map { String(format: "%02x", $0) }.joined()
        let reply = try await post("v1/register", ["ticket": ticket, "platform": "apns", "token": token])
        guard let reference = reply["reference"] as? String, Self.valid(reference),
            let management = reply["management_token"] as? String, Self.valid(management),
            let expiry = reply["expires"] as? NSNumber else { throw GComsError.invalidResponse }
        var state = try await storage.load()
        state.reference = reference
        state.managementToken = management
        state.expires = expiry.uint64Value
        try await storage.save(state)
    }
    /// Persist a revision before binding; repeat after reopen and before suspension.
    public func bind(to sdk: GComs) async throws {
        guard !busy else { throw GComsError.unavailable }
        busy = true
        defer { busy = false }
        var state = try await storage.load()
        let reference = state.reference ?? String(repeating: "0", count: 64)
        guard Self.valid(reference), state.revision < UInt64.max else { throw GComsError.invalidResponse }
        let now = UInt64(Date().timeIntervalSince1970)
        let expiry = state.reference == nil ? now + 3600 : min(now + 23 * 3600, state.expires)
        guard expiry > now else { throw GComsError.native("Refresh expired push registration first") }
        state.revision += 1
        try await storage.save(state)
        let chars = Array(reference)
        let bytes = stride(from: 0, to: 64, by: 2).map { UInt8(String(chars[$0..<$0 + 2]), radix: 16)! }
        _ = try await sdk.request(JSONSerialization.data(withJSONObject: [
            "op": "bind_push", "reference": bytes, "revision": state.revision, "expires": expiry
        ]))
    }
    /// Disable gateway delivery, then call bind(to:) to remove current relay bindings.
    public func unregister() async throws {
        guard !busy else { throw GComsError.unavailable }
        busy = true
        defer { busy = false }
        var state = try await storage.load()
        if let reference = state.reference, let token = state.managementToken {
            _ = try await post("v1/unregister", ["reference": reference, "management_token": token])
        }
        state.reference = nil
        state.managementToken = nil
        state.expires = 0
        try await storage.save(state)
    }
    public nonisolated static func hintReference(_ userInfo: [AnyHashable: Any]) -> String? {
        guard userInfo["gcoms_activity"] as? String == "message",
            let reference = userInfo["gcoms_reference"] as? String, valid(reference),
            Set(userInfo.keys.compactMap { $0 as? String }) == Set(["aps", "gcoms_activity", "gcoms_reference"]) else { return nil }
        return reference
    }
    private nonisolated static func valid(_ value: String) -> Bool {
        value.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil
    }
}

private final class NoRedirect: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
}

/// Non-synchronizing device-only storage. One instance per profile.
public actor KeychainPushStorage: PushStorage {
    private let account: String
    public init(profile: String) throws {
        guard profile.range(of: "^[A-Za-z0-9_-]{1,64}$", options: .regularExpression) != nil else {
            throw GComsError.invalidResponse
        }
        account = profile
    }
    private var query: [String: Any] {
        [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: "boo.gcoms.push",
         kSecAttrAccount as String: account, kSecAttrSynchronizable as String: false]
    }
    public func load() throws -> PushState {
        var request = query
        request[kSecReturnData as String] = true
        request[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: CFTypeRef?
        let status = SecItemCopyMatching(request as CFDictionary, &result)
        if status == errSecItemNotFound { return PushState() }
        guard status == errSecSuccess, let data = result as? Data, data.count <= 8192 else {
            throw GComsError.native("Push keychain is unavailable")
        }
        return try JSONDecoder().decode(PushState.self, from: data)
    }
    public func save(_ value: PushState) throws {
        let data = try JSONEncoder().encode(value)
        guard data.count <= 8192 else { throw GComsError.invalidResponse }
        let attributes: [String: Any] = [kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly]
        var status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if status == errSecItemNotFound {
            status = SecItemAdd(query.merging(attributes) { _, value in value } as CFDictionary, nil)
        }
        guard status == errSecSuccess else { throw GComsError.native("Push keychain write failed") }
    }
}
